Source/DafnyRuntime/DafnyRuntime.go

```diff
@@ -692,183 +692,412 @@ func (seq Seq) VerbatimString(asLiteral bool) string {
-  ans := newArray(dims...)
-  if init != nil {
-    for i := range ans.contents {
-      ans.contents[i] = init
+// newArrayWithValues returns a new one-dimensional Array with the given initial
+// values. It is only used internally, by *Builder.ToArray().
+func newArrayWithValues(values ...interface{}) Array {
+  totalLength := len(values)
+  intDims := []int{totalLength}
+  if totalLength == 0 {
+    return newZeroLengthArray(intDims)
+  }
+
+  // Inspect the type of "values[0]" to consider Array specialization
+  if _, ok := values[0].(byte); ok {
+    arr := make([]byte, totalLength)
+    for i := range arr {
+      arr[i] = values[i].(byte)
+    }
+    return &ArrayForByte{
+      contents: arr,
+      dims:     intDims,
+    }
+  }
+  if _, ok := values[0].(Char); ok {
+    arr := make([]Char, totalLength)
+    for i := range arr {
+      arr[i] = values[i].(Char)
+    }
+    return &ArrayForChar{
+      contents: arr,
+      dims:     intDims,
     }
   }
-  return ans
-}
 
-// NewArrayWithValues returns a new one-dimensional Array with the given initial
-// values.
-func NewArrayWithValues(values ...interface{}) *Array {
-  arr := make([]interface{}, len(values))
+  // Use the default representation
+  arr := make([]interface{}, totalLength)
   copy(arr, values)
-  return &Array{
+  return &ArrayStruct{
     contents: arr,
-    dims:     []int{len(values)},
+    dims:     intDims,
   }
 }
 
-// Len returns the length of the array in the given dimension.
-func (array *Array) Len(dim int) Int {
-  return IntOf(array.LenInt(dim))
+// NewArrayWithValue returns a new Array full of the given initial value.
+func NewArrayWithValue(init interface{}, dims ...Int) Array {
+  return NewArrayFromExample(init, init, dims...)
 }
 
-// LenInt returns the length of the array in the given dimension, as an int.
-func (array *Array) LenInt(dim int) int {
-  return array.dims[dim]
+// NewArray returns a new Array full of the default value of the given type.
+func NewArray(dims ...Int) Array {
+  return NewArrayFromExample(nil, nil, dims...)
 }
 
-// Equals compares two arrays for equality.  Values are compared using
-// dafny.AreEqual.
-func (array *Array) Equals(array2 *Array) bool {
-  if array == array2 {
-    return true
+/***** ArrayStruct is default implementation of the Array interface. *****/
+
+type ArrayStruct struct {
+  contents []interface{} // stored as a flat one-dimensional slice
+  dims     []int
+}
+
+func (_this ArrayStruct) dimensionCount() int {
+  return len(_this.dims)
+}
+
+func (_this ArrayStruct) dimensionLength(dim int) int {
+  return _this.dims[dim]
+}
+
+func (_this ArrayStruct) ArrayGet1(index int) interface{} {
+  return _this.contents[index]
+}
+
+func (_this ArrayStruct) ArraySet1(value interface{}, index int) {
+  _this.contents[index] = value
+}
+
+func (_this ArrayStruct) ArrayGet1Byte(index int) byte {
+  panic("cannot specialize general ArrayGet with byte")
```

Comment:
```suggestion
  panic("Expected specialized array type that contains bytes, but found general-purpose array of interface{}")
```

Users should never see this but I think that message helps point debuggers towards what happened better. Similar edits to the other error messages.

An alternative would be to actually downcast to byte, but I do think it's better to panic if this should never happen.